Step3

Sehun-Kim opened the pull request:
### image 파일 업로드 및 게임 순위 표시 기능

- 회원가입시 이미지를 서버에 업로드 해주는 `ImageUploader`
- `User`의 사진 정보(이름, 저장 위치)를 묶은 @Imbeddable 클래스 `Picture`
- 회원가입시 이미지를 업로드 하지 않는다면 `User`는 기본 이미지를 갖는다.
- 메인 화면에 승리횟수(winCnt)를 기준으로 상위 10명을 표시함

Commit:
image upload

이미지 업로드 기능 완성
- imageUploadservice : 이미지 업로드 기능을 담당.
- Picture class : db에 담길 이미지정보를 가진 엔티티

Commit:
rank

winCnt를 기준으로 높은 10명을 표시
- 각user의 이미지가 보이도록함
- 가입시 이미지를 저장하지 않으면 기본 이미지가 표시됨

javajigi commented on src/main/java/indianpoker/domain/user/Picture.java:
```diff
@@ -0,0 +1,36 @@
+
+@Embeddable
+public class Picture {
+    private static final String DEFAULT_PICTURE_NAME = "example.jpeg";
```
이런 기본 값을 application.properties에서 관리하고 @Value 애노테이션으로 가져오도록 구현해 본다.
https://www.baeldung.com/spring-value-annotation 문서 참고

javajigi commented on src/main/java/support/domain/ImageUploader.java:
```diff
@@ -0,0 +1,63 @@
+        destinationFile.getParentFile().mkdirs();
+        pic.transferTo(destinationFile);
+
+        return new Picture(fileName, destinationFile.getPath());
```
`destinationFile.getPath()` 정보를 가지고 있을 경우 추후 "String uploadFileDir" 정보가 변경되는 경우 문제가 될 수 있지 않을까?

javajigi commented on src/main/java/indianpoker/domain/user/Picture.java:
```diff
@@ -0,0 +1,36 @@
+    private static final String DEFAULT_PICTURE_NAME = "example.jpeg";
+
+    @Column(unique = true)
+    private String fileName;
```
fileName의 유일성을 db를 통하지 않고 애플리케이션 코드에서 처리하면 어떨까?
UUID와 같은 녀석을 사용해 파일 이름이 충돌하지 않도록 하면 어떨까?

javajigi reviewed (changes requested):
구현 잘 했네. 👍 
피드백 남겼어요.

Commit:
feedback reflect

image 저장소를 프로젝트가 아닌 외부 폴더로 옮김
- 이미지 확장자의 유효성 검사를 Picture에게 위임
image 파일을 불로올 수 있는 imageController 생성
- 이미지 이름을 받아 외부폴더에서 읽어온 이미지 리소스를 반환해줌
이미지의 중복된 이름을 피하기 위해 UUID 사용
- Picture는 확장자를 따로 저장하여 경로를 만들 때 uuid와 확장자를 합쳐서 경로를리턴해줌
